src/hex2nix.erl

```diff
@@ -0,0 +1,199 @@
+
+-spec split_data_into_versions_and_detail([term()]) ->
+                                                 {[app()], [{app(), app_detail()}]}.
+split_data_into_versions_and_detail(AppData) ->
+    lists:partition(fun({Name, _}) when erlang:is_binary(Name) ->
+                            true;
+                       (_) ->
+                            false
+                    end, AppData).
+
+-spec find_all_buildable_versions(dict:dict(app(), app_detail())) ->
+                                         dict:dict(app_name(), [app_version()]).
+find_all_buildable_versions(AppData) ->
+    dict:fold(fun({Name, Version}, _, Acc) ->
+                      dict:append(Name, Version, Acc)
+                end, dict:new(), AppData).
+
+-spec cleanup_app_data([any()]) -> dict:dict(app(), app_detail()).
+cleanup_app_data(AppData) ->
+    lists:foldl(fun({App, [Deps, _, BuildSystems]}, Acc) ->
+                        case is_supported_build_system(BuildSystems) of
+                            true ->
+                                dict:store(App, simplify_deps(Deps), Acc);
+                            false ->
+                                Acc
+                        end;
+                   (X, Acc) ->
+                        io:format("Discarding malformed detail: ~p~n", [X]),
+                        Acc
+                end, dict:new(), AppData).
+
+-spec decompress_file(file:filename(), file:filename()) ->
+                             {ok, file:filename()}.
+decompress_file(GzippedFileName, TargetDirectory) ->
+    ResultFile = filename:join(TargetDirectory, "registry.ets"),
+    {ok, Data} = file:read_file(GzippedFileName),
+    UncompressedData = zlib:gunzip(Data),
+    file:write_file(ResultFile, UncompressedData, []),
+    io:format("File uncompressed to ~s~n", [ResultFile]),
+    {ok, ResultFile}.
+
+-spec simplify_deps([[any()]]) ->
+                           [app()].
+simplify_deps(Deps) ->
+    lists:map(fun simplify_dep/1, Deps).
+
+-spec simplify_dep([app_name() | binary()]) ->
+                          {app_name(), [binary()]}.
+simplify_dep([Name, Constraint, _, _]) ->
+    {Name, Constraint}.
+
+-spec is_supported_build_system([binary()] | binary()) -> boolean().
+is_supported_build_system(SystemList)
+  when erlang:is_list(SystemList) ->
+    lists:any(fun is_supported_build_system/1, SystemList);
+is_supported_build_system(<<"rebar3">>) -> true;
+is_supported_build_system(<<"make">>) ->  true;
+is_supported_build_system(<<"rebar">>) ->  true;
+is_supported_build_system(_) ->  false.
+
+-spec reduce_to_latest_buildable_version(dict:dict(app(), app_detail()),
+                                         [{app_name(), [app_version()]}]) ->
+                                                [app()].
+reduce_to_latest_buildable_version(AppData, AppVersions) ->
+    lists:foldl(fun({Name, [Versions]}, Acc) ->
+                        [HardVersion | _] = lists:sort(fun ec_semver:gte/2, Versions),
```

Comment:
Hard version is correct. I meant a resolved version, a 'hard' version. but 'head' version might work too.